tools/PI/DevHome.PI/Models/WERAnalyzer.cs

```diff
@@ -0,0 +1,184 @@
+                {
+                    _registeredAnalysisTools.Add(tool);
+                }
+            }
+
+            foreach (var report in _werAnalysisReports)
+            {
+                RunToolAnalysis(report, e.NewItems.Cast<Tool>().ToList());
+            }
+        }
+
+        // Or if we removed a tool, remove the analysis that it did.
+        else if (e.Action == NotifyCollectionChangedAction.Remove && e.OldItems is not null)
+        {
+            foreach (Tool tool in e.OldItems)
+            {
+                if (tool.Type.HasFlag(ToolType.DumpAnalyzer))
+                {
+                    _registeredAnalysisTools.Remove(tool);
+                }
+
+                foreach (var report in _werAnalysisReports)
+                {
+                    report.RemoveToolAnalysis(tool);
+                }
+            }
+        }
+    }
+
+    private void WER_CollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
+    {
+        if (e.Action == NotifyCollectionChangedAction.Add && e.NewItems != null)
+        {
+            ProcessDumpList(e.NewItems.Cast<WERReport>().ToList());
+        }
+    }
+
+    private void PopulateCurrentLogs()
+    {
+        ProcessDumpList(_werHelper.WERReports.ToList<WERReport>());
+    }
+
+    private void ProcessDumpList(List<WERReport> reports)
+    {
+        List<WERAnalysisReport> reportsToAnalyze = new();
+
+        // First publish all of these reports to our listeners. Then we'll go back and perform
+        // analysis on them.
+        foreach (var report in reports)
+        {
+            var reportAnalysis = new WERAnalysisReport(report);
+
+            _werAnalysisReports.Add(reportAnalysis);
+            reportsToAnalyze.Add(reportAnalysis);
+
+            // When the crash dump path changes, we'll want to perform analysis on it.
+            report.PropertyChanged += (sender, e) =>
+            {
+                if (e.PropertyName == nameof(WERReport.CrashDumpPath))
+                {
+                    RunToolAnalysis(reportAnalysis, RegisteredAnalysisTools.ToList<Tool>());
+                }
+            };
+        }
+
+        List<Tool> tools = RegisteredAnalysisTools.ToList<Tool>();
+        foreach (var reportAnalysis in reportsToAnalyze)
+        {
+            RunToolAnalysis(reportAnalysis, tools);
+        }
+    }
+
+    private void RunToolAnalysis(WERAnalysisReport report, List<Tool> tools)
+    {
+        if (string.IsNullOrEmpty(report.Report.CrashDumpPath))
+        {
+            // We need a crash dump to perform an analysis
+            return;
+        }
+
+        foreach (Tool tool in tools)
+        {
+            AnalysisRequest request = new()
+            {
+                Report = report,
+                Tool = tool,
+            };
+
+            // Queue the request that will be processed on a separate thread
+            _analysisRequests.Add(request);
+        }
+    }
+
+    public void Dispose()
```

Comment:
I was going to comment on unsubscribing events here to prevent trapping objects, but this class (the subscriber) seems to be a singleton.

Reply:
Yep.